### PR TITLE
Add new property in storagedeviceset

### PR DESCRIPTION
**Pull request by crombus**
add new property in storagedeviceset

DeviceType is the property which will allow admin to add the device type in Storagecluster CR.

### CODE REVIEW

**Review by nbalacha — CHANGES_REQUESTED** [review]
There is a typo in the commit message - "stroageclusterdeviceset"
I also think it would be better to say storageDeviceSet instead of storageClusterDeviceSet (which does not exist).

File: pkg/controller/storagecluster/reconcile.go [hunk 1]
```diff
@@ -406,6 +406,11 @@ func (r *ReconcileStorageCluster) validateStorageDeviceSets(sc *ocsv1.StorageClu
 				return fmt.Errorf("failed to validate StorageDeviceSet %d: no StorageClass specified for metadataPVCTemplate", i)
 			}
 		}
+		if ds.DeviceType != "" {
+			if ds.DeviceType == "SSD" || ds.DeviceType == "HDD" || ds.DeviceType == "NVMe" {
```
Comment (vbnrh) [inline pkg/controller/storagecluster/reconcile.go:410]: ignore the case.

**Review by vbnrh — CHANGES_REQUESTED** [review]

File: deploy/crds/ocs.openshift.io_storageclusters_crd.yaml [hunk 2]
```diff
@@ -1447,6 +1447,9 @@ spec:
                           phase:
                             description: Phase represents the current phase of PersistentVolumeClaim.
                             type: string
+                  devicetype:
+                    description: DeviceType is the value of device type in this StorageDeviceSet
```
Comment (vbnrh) [inline deploy/crds/ocs.openshift.io_storageclusters_crd.yaml:1451]: DeviceType represents the value of underlying storage device used (restricted to 3 values; SSD, NVME, HDD)

File: deploy/csv-templates/crds/ocs.openshift.io_storageclusters_crd.yaml [hunk 3]
```diff
@@ -1447,6 +1447,9 @@ spec:
                           phase:
                             description: Phase represents the current phase of PersistentVolumeClaim.
                             type: string
+                  devicetype:
+                    description: DeviceType is the value of device type in this StorageDeviceSet
```
Comment (vbnrh) [inline deploy/csv-templates/crds/ocs.openshift.io_storageclusters_crd.yaml:1451]: DeviceType represents the value of underlying storage device used (restricted to 3 values; SSD, NVME, HDD)

File: deploy/olm-catalog/ocs-operator/manifests/storagecluster.crd.yaml [hunk 4]
```diff
@@ -1440,6 +1440,9 @@ spec:
                             type: string
                         type: object
                     type: object
+                  devicetype:
+                    description: DeviceType is the value of device type in this StorageDeviceSet
```
Comment (vbnrh) [inline deploy/olm-catalog/ocs-operator/manifests/storagecluster.crd.yaml:1444]: DeviceType represents the value of underlying storage device used (restricted to 3 values; SSD, NVME, HDD)

**Review by vbnrh — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/reconcile.go [hunk 5]
```diff
@@ -406,6 +407,11 @@ func (r *ReconcileStorageCluster) validateStorageDeviceSets(sc *ocsv1.StorageClu
 				return fmt.Errorf("failed to validate StorageDeviceSet %d: no StorageClass specified for metadataPVCTemplate", i)
 			}
 		}
+		if ds.DeviceType != "" {
+			if strings.EqualFold(ds.DeviceType, "SSD") || strings.EqualFold(ds.DeviceType, "HDD") || strings.EqualFold(ds.DeviceType, "NVMe") {
```
Comment (iamniting) [inline pkg/controller/storagecluster/reconcile.go:411]: It looks like you wanted to use a not condition here as per the error msg.

Comment (iamniting) [inline pkg/controller/storagecluster/reconcile.go:411]: It is good to define these values as types and use it here

**Review by iamniting — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/reconcile.go (hunk 5 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go:411]: yes. I'll change that.

Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go:411]: can you explain more what types and how can we ensure to ignore case insensitivity.

Comment (iamniting) [reply pkg/controller/storagecluster/reconcile.go]: something like this https://github.com/openshift/ocs-operator/blob/master/pkg/controller/storagecluster/uninstall_reconciler.go#L24-L43

Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go]: why would I want to use this approach. when I want to use devicetype only once in file. it is not like `ssd` or `ssd` is being used more than once.

Comment (iamniting) [reply pkg/controller/storagecluster/reconcile.go]: It was just a suggestion and totally up to you.

File: pkg/controller/storagecluster/reconcile.go [hunk 6]
```diff
@@ -406,6 +407,11 @@ func (r *ReconcileStorageCluster) validateStorageDeviceSets(sc *ocsv1.StorageClu
 				return fmt.Errorf("failed to validate StorageDeviceSet %d: no StorageClass specified for metadataPVCTemplate", i)
 			}
 		}
+		if ds.DeviceType != "" {
+			if !strings.EqualFold(ds.DeviceType, "SSD") || !strings.EqualFold(ds.DeviceType, "HDD") || !strings.EqualFold(ds.DeviceType, "NVMe") {
```
Comment (jarrpa) [inline pkg/controller/storagecluster/reconcile.go:411]: Don't just compare strings here. Set them as consts and create an array of valid types. Also, are they actually case-sensitive/mixed case, or can they be all lowercase?

**Review by jarrpa — CHANGES_REQUESTED** [review]
You check the value od `ds.DeviceType` but then you don't seem to do anything with it. What's the point of this PR?

Also, please split the CSV changes into a separate commit.

File: pkg/controller/storagecluster/reconcile.go (hunk 5 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go]: I would make it a demand. :) It is good practice to use explicit types rather than just comparing raw strings. You could then also create an array of these types, and just search to see if the array contains the given value.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 7]
```diff
@@ -106,6 +106,12 @@ type StorageDeviceSet struct {
 	// +optional
 	Replica int `json:"replica,omitempty"`
 
+	// DeviceType is the value of device type in
+	// this StorageDeviceSet. It can have only
+	// three values (SSD, HDD, NVMe)
+	// +optional
+	DeviceType string `json:"devicetype,omitempty"`
```
Comment (umangachapagain) [inline pkg/apis/ocs/v1/storagecluster_types.go:113]: If these are the only possible values we should be able to error out immediately if any other value is passed. Can you look into adding that?

Comment (umangachapagain) [inline pkg/apis/ocs/v1/storagecluster_types.go:113]:
```suggestion
	DeviceType string `json:"deviceType,omitempty"`
```

File: pkg/controller/storagecluster/reconcile.go [hunk 8]
```diff
@@ -406,6 +407,11 @@ func (r *ReconcileStorageCluster) validateStorageDeviceSets(sc *ocsv1.StorageClu
...
+				return fmt.Errorf("failed to validate DeviceType %s: no Device of this type", ds.DeviceType)
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:412]:
```suggestion
				return fmt.Errorf("failed to validate DeviceType %q: no device of this type", ds.DeviceType)
```

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/reconcile.go (hunk 6 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go:411]: they all can be lower case also. mixed case can be for only NVMe. It is just a logical grouping which will added as a metadata for osd.

File: pkg/controller/storagecluster/reconcile.go (hunk 5 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go]: the main concern is to check for case sensitivity. https://github.com/openshift/ocs-operator/pull/859#discussion_r513376637

**Comment by crombus** [comment]
After checking ideally it the device type should be in upper case or lower case an array can be made with 6 values(upper case and lower case). Then I can check for `ds.DeviceType` contain in the string. @jarrpa @iamniting

File: pkg/apis/ocs/v1/storagecluster_types.go (hunk 7 above)
Comment (crombus) [reply pkg/apis/ocs/v1/storagecluster_types.go:113]: there can be case sensitive case. https://github.com/openshift/ocs-operator/pull/859#issuecomment-717880007

**Comment by nbalacha** [comment]
> add new property in stroagedeviceset
> DeviceType

Nit: Typo. 

"stroagedeviceset" should be "storagedeviceset"

**Comment by crombus** [comment]
> After checking ideally it the device type should be in upper case or lower case an array can be made with 6 values(upper case and lower case). Then I can check for `ds.DeviceType` contain in the string. @jarrpa @iamniting

NVMe is  abbreviated like this I think adding this in the array too.

File: pkg/controller/storagecluster/reconcile.go (hunk 6 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:411]: If that's the case then you don't nee dto codify the multiple case types. Just force the string to always be lowercase and compare those values.

Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go:411]: ok. I'll update the PR

File: pkg/apis/ocs/v1/storagecluster_types.go (hunk 7 above)
Comment (jarrpa) [reply pkg/apis/ocs/v1/storagecluster_types.go:113]: This still needs to be addressed

File: pkg/controller/storagecluster/reconcile.go (hunk 6 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:412]: You have not finished the requested changes. Please define the strings as consts, preferably with a new `type DeviceType string`.

File: pkg/controller/storagecluster/reconcile.go (hunk 8 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:413]: This still needs to be addressed.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 9]
```diff
@@ -43,7 +43,7 @@ type StorageClusterSpec struct {
 	// If enabled, sets the failureDomain to host, allowing devices to be
 	// distributed evenly across all nodes, regardless of distribution in zones
 	// or racks.
-	FlexibleScaling  bool     `json:"flexibleScaling,omitempty"`
+	FlexibleScaling bool `json:"flexibleScaling,omitempty"`
```
Comment (jarrpa) [inline pkg/apis/ocs/v1/storagecluster_types.go:46]: This does not belong in this PR, please remove it.

**Review by jarrpa — CHANGES_REQUESTED** [review]
Multiple review comments remain unaddressed. In addition, the title of this PR is vague and inadequate. Please change the subject to "Add DeviceType to StorageDeviceSet" for *both* the PR and the commit.

**Comment by crombus** [comment]
@jarrpa one the test for crd verification failing. how to solve that?

File: pkg/controller/storagecluster/reconcile.go [hunk 10]
```diff
@@ -33,6 +34,9 @@ type ReconcileStrategy string
 // StorageClassProvisionerType is a string representing StorageClass Provisioner. E.g: aws-ebs
 type StorageClassProvisionerType string
 
+// DeviceTypes is list of valid device type from which storagecluster can be created
+var DeviceTypes = []string{"nvme", "ssd", "hdd"}
```
Comment (iamniting) [inline pkg/controller/storagecluster/reconcile.go:38]: This is still not done, make them constants as per https://github.com/openshift/ocs-operator/pull/859

**Review by iamniting — CHANGES_REQUESTED** [review]

**Comment by jarrpa** [comment]
> @jarrpa one the test for crd verification failing. how to solve that?

Remove the changes to the CRD validation test, as they are invalid.

File: pkg/controller/storagecluster/reconcile.go (hunk 10 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go:38]: I went with this approach https://github.com/openshift/ocs-operator/pull/859#discussion_r516121654.  Not against of adding const it is just that there will be devices which will be added in future as of now PCI will be one. creating const for each one. and array of devices which will better? and there will one benefit of using const that will be when we will tune the osd. So i am like 5050 on both approaches .

Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:38]: I am 100% for consts, so since you're not fully against it I'd say go ahead with the change. :)

**Comment by nbalacha** [comment]
> add new property in stroagedeviceset
> DeviceType

The typo still exists.

Please add more detail as to the use case for this - when is this property used?

File: pkg/controller/storagecluster/reconcile.go (hunk 10 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go]: consts it is. Added.

**Comment by jarrpa** [comment]
Please run `make ocs-operator-ci` before pushing to GitHub. :)

File: pkg/controller/storagecluster/reconcile.go [hunk 11]
```diff
@@ -33,6 +34,9 @@ type ReconcileStrategy string
 // StorageClassProvisionerType is a string representing StorageClass Provisioner. E.g: aws-ebs
 type StorageClassProvisionerType string
 
+// DeviceTypes is a string representing DeviceType
```
Comment (jarrpa) [inline pkg/controller/storagecluster/reconcile.go:37]: New structs shouldn't be plural.

```suggestion
// DeviceType is a string representing a storage device type
```

**Review by jarrpa — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/reconcile.go (hunk 11 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go]: done

File: test/crd_validation_test.go [hunk 12]
```diff
@@ -36,7 +36,6 @@ func TestSampleCustomResources(t *testing.T) {
 	root := "./../deploy/crds"
 	crdCrMap := map[string]string{
 		"ocs.openshift.io_ocsinitializations_crd.yaml": "ocs_v1_ocsinitialization_cr",
-		"ocs.openshift.io_storageclusters_crd.yaml":    "ocs_v1_storagecluster_cr",
```
Comment (jarrpa) [inline test/crd_validation_test.go:39]: This is not correct. You want to maintain the validation of StorageCluster, you just have to change the validation parameters.

**Review by jarrpa — CHANGES_REQUESTED** [review]
Code changes look good, but the change in the test is invalid.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 13]
```diff
@@ -106,6 +106,12 @@ type StorageDeviceSet struct {
 	// +optional
 	Replica int `json:"replica,omitempty"`
 
+	// DeviceType is the value of device type in
+	// this StorageDeviceSet. It can have only
+	// three values (SSD, HDD, NVMe)
+	// +optional
+	DeviceType string `json:"deviceType,omitempty"`
```
Comment (umangachapagain) [inline pkg/apis/ocs/v1/storagecluster_types.go:113]:
```suggestion
	// DeviceType is the value of device type in
	// this StorageDeviceSet. It can have only
	// three values (SSD, HDD, NVMe)
	// +kubebuilder:validation:Enum=SSD;ssd;HDD;hdd;NVMe;NVME;nvme
	// +optional
	DeviceType string `json:"deviceType,omitempty"`
```

This will validate user input at client side.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 14]
```diff
@@ -106,6 +106,12 @@ type StorageDeviceSet struct {
 	// +optional
 	Replica int `json:"replica,omitempty"`
 
+	// DeviceType is the value of device type in
+	// this StorageDeviceSet. It can have only
+	// three values (SSD, HDD, NVMe)
```
Comment (umangachapagain) [inline pkg/apis/ocs/v1/storagecluster_types.go:111]:
```suggestion
	// this StorageDeviceSet. It can have one of
	// three values (SSD, HDD, NVMe)
```

File: pkg/controller/storagecluster/reconcile.go [hunk 15]
```diff
@@ -62,6 +66,15 @@ osd_memory_target_cgroup_limit_ratio = 0.5
...
 	ReconcileStrategyStandalone ReconcileStrategy = "standalone"
+
+	// DeviceTypeSSD represents the DeviceType SSD
+	DeviceTypeSSD DeviceType = "ssd"
+
+	// DeviceTypeHDD represents the DeviceType HDD
+	DeviceTypeHDD DeviceType = "hdd"
+
+	// DeviceTypeNVMe represents the DeviceType NVMe
+	DeviceTypeNVMe DeviceType = "nvme"
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:77]: Since the variables have `DeviceType` prefix, I don't see the benefit for using `DeviceType` alias for `string` type.

We can either do `SSD DeviceType = "ssd"` or `DeviceTypeSSD string = "ssd"`

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/reconcile.go (hunk 15 above)
Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go:77]: I did this because it has been done similar for adding const and naming just SSD or HDD fails for golint test

Comment (umangachapagain) [reply pkg/controller/storagecluster/reconcile.go:77]: Like in Line 44 53 57,  `DeviceTypeSSD = "ssd"` should be enough too.

Comment (crombus) [reply pkg/controller/storagecluster/reconcile.go]: done

**Comment by jarrpa** [comment]
CSV test failures seem valid, please investigate. Looks like it may need a rebase?

**Review by jarrpa — APPROVED** [review]
/lgtm
